classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $retries = $record->retries + 1;
+        // Update retries.
+        $this->update_retries_in_table($record->id, $retries);
+        $logs = [];
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log) = $mut->merge($record->keepuserid, $record->removeuserid);
+        if (!$success) {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        } else {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        }
+        // Append logs to the list.
+        // $baseitem->log[$item->timemodified] = json_decode($item->log, false);
+        $logs[$record->timeadded] = $log;
+        $this->update_status_and_log_in_table($record->id,
+                                            $status,
+                                            $logs);
+        $record->log = $logs;
+        $record->status = $status;
+        return $record;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log_in_table(int $idrecord,
+                                                    int $status,
+                                                    array $log): void {
+        global $DB;
+        if ($status == merge_request::COMPLETED_WITH_SUCCESS ||
+                    $status == merge_request::COMPLETED_WITH_ERRORS) {
+            $update = (object)[
+                'id' => $idrecord,
+                'status' => $status,
+                'timecompleted' => time(),
+                'log' => json_encode($log),
+            ];
+        } else {
+            $update = (object)[
+                'id' => $idrecord,
+                'status' => $status,
+                'log' => json_encode($log),
+            ];
+        };
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            $update
+        );
+    }
+    /**
+     * Function for verifying the users to keep and remove.
+     */
+    private function verify_users_to_keep_and_remove(object $record): void {
+        global $DB;
+        $keepuserfield = $record->keepuserfield;
+        $keepuservalue = $record->keepuservalue;
+        $removeuserfield = $record->removeuserfield;
+        $removeuservalue = $record->removeuservalue;
+        // Verify user to remove.
+        $usertoremove = $DB->get_records(merge_request::TABLE_USERS, [$removeuserfield => $removeuservalue]);
```

Comment:
Unnecessary line: `$usertoremove` is never used and this line is already executed in the `find_user_id_or_fails`.

Reply:
removed